File: Altf4/backend/src/backend.py
"""
Documentazione dell'app ALT+F4
----------------------------------
!!! ATTENZIONE !!! 
PER USARE QUESTO SCRIPT DOVETE INSTALLARE MariaDB, vi metto il link cosi tutti usiamo lo stesso,

https://dlm.mariadb.com/4257323/MariaDB/mariadb-11.4.6/winx64-packages/mariadb-11.4.6-winx64.msi

durante l'installazione inserite la password quando ve la chiede e metteteci 123456, non vi dovrebbe
far impostare l'username se siete da windows, comunque lasciate root.
Creare le cartelle come vi mandero in foto che almeno ce le abbiamo tutti uguali, non mi va di setuppare
il git hub se no avremmo gia' risolto il problema
Per ora l'ho testata e funziona, se non runna state facendo qualcosa di sbagliato, chiedete a chat,
particolare attenzione al file data.tsv, che vi allego su discord, perche' deve stare all'interno della
cartella backend
!!!FINE ATTENZIONE!!!

backend.py
effettua la connessione a Mariadb
popola il database altf4 da data.tsv gestendo errori
funzioni:-inserisci team -inserisce team nel database se non presente
         -stampa partite -stampa tutte le partite all'interno del database utile per vedere se il database e' stato popolato correttamente

         
P.S fatemi sapere se vi piace la documentazione scritta cosi', da quel che ho capito il bro ci fa caso a come la scriviamo
"""
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
import json
import csv
import mariadb
import sys

# Configurazione della connessione
try:
    conn = mariadb.connect(
        user="root",
        password="123456",
        host="localhost",
        port=3306,
        database="altf4"
    )
except mariadb.Error as e:
    print(f"Errore nella connessione al database: {e}")
    sys.exit(1)

cursor = conn.cursor()

# Funzione per inserire un team se non esiste
def inserisci_team(nome_team):
    try:
        cursor.execute("INSERT IGNORE INTO team (nome) VALUES (?)", (nome_team,))
    except mariadb.Error as e:
        print(f"Errore nell'inserimento del team {nome_team}: {e}")

def stampa_partite():
    try:
        conn = mariadb.connect(
            user="root",
            password="123456",
            host="localhost",
            port=3306,
            database="altf4"
        )
        cursor = conn.cursor()

        cursor.execute("""
            SELECT team_alt, team_enemy, data, ora, risultato
            FROM partita
            ORDER BY data, ora
        """)

        partite = cursor.fetchall()

        print("\nPartite registrate:\n")
        for p in partite:
            team_alt, team_enemy, data, ora, risultato = p
            risultato = risultato or "—"
            print(f"{data} {ora} — {team_alt} vs {team_enemy} → {risultato}")

        cursor.close()
        conn.close()

    except mariadb.Error as e:
        print(f"Errore durante la connessione o la query: {e}")


# Lettura del file TSV e inserimento dei dati
with open('data.tsv', 'r', encoding='utf-8') as file_tsv:
    lettore = csv.reader(file_tsv, delimiter='\t')
    next(lettore)  # salta intestazione

    for riga in lettore:
        if len(riga) != 5:
            print(f"Riga malformata: {riga}")
            continue

        team_alt, team_enemy, data_partita, ora_partita, risultato = riga

        # Converti la data da gg/mm/aaaa a aaaa-mm-gg
        try:
            giorno, mese, anno = data_partita.strip().split('/')
            data_partita = f"{anno}-{mese}-{giorno}"
        except ValueError:
            print(f"Formato data non valido: {data_partita}")
            continue

        # Inserimento dei team
        inserisci_team(team_alt)
        inserisci_team(team_enemy)

        # Inserimento della partita
        try:
            cursor.execute("""
                INSERT INTO partita (team_alt, team_enemy, data, ora, risultato)
                VALUES (?, ?, ?, ?, ?)
            """, (team_alt, team_enemy, data_partita, ora_partita, risultato if risultato else None))
        except mariadb.Error as e:
            print(f"Errore nell'inserimento della partita: {e}")

# Salvataggio delle modifiche e chiusura della connessione
conn.commit()
cursor.close()
conn.close()

app = FastAPI()

# Abilita richieste da localhost:3000 (React)
app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:5173"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

@app.get("/partite")
def get_partite():
    try:
        conn = mariadb.connect(
            user="root",
            password="123456",
            host="localhost",
            port=3306,
            database="altf4"
        )
        cursor = conn.cursor()

        cursor.execute("""
            SELECT team_alt, team_enemy, data, ora, risultato
            FROM partita
            ORDER BY data, ora
        """)

        rows = cursor.fetchall()

        result = [
            {
                "team_alt": r[0],
                "team_enemy": r[1],
                "data": r[2].isoformat(),
                "ora": str(r[3])[:-3],  # "18:00:00" → "18:00"
                "risultato": r[4]
            }
            for r in rows
        ]

        cursor.close()
        conn.close()
        return result

    except mariadb.Error as e:
        print(f"Errore in /partite: {e}")
        return {"errore": str(e)}

File: Altf4/backend/src/backend.js
import express from 'express';
import cors from 'cors';
import fs from 'fs';
import mariadb from 'mariadb';
import path from 'path';
import { fileURLToPath } from 'url';

// Risolvi __dirname in ambiente ES Modules
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configurazione del server Express
const app = express();
app.use(cors({ origin: 'http://localhost:5173' }));  // Permette richieste da React

// Configurazione del pool di connessioni a MariaDB
const pool = mariadb.createPool({
  host: 'localhost',
  port: 3306,
  user: 'root',
  password: '123456',
  database: 'altf4',
  connectionLimit: 5,
});

// Funzione per resettare le tabelle e popolare il database da data.tsv all'avvio
async function initDb() {
  let conn;
  try {
    conn = await pool.getConnection();

    // Drop e ricreazione delle tabelle
    await conn.query('DROP TABLE IF EXISTS partita');
    await conn.query('DROP TABLE IF EXISTS team');
    await conn.query(`
      CREATE TABLE team (
        id INT AUTO_INCREMENT PRIMARY KEY,
        nome VARCHAR(255) NOT NULL UNIQUE
      )
    `);
    await conn.query(`
      CREATE TABLE partita (
        id INT AUTO_INCREMENT PRIMARY KEY,
        team_alt VARCHAR(255) NOT NULL,
        team_enemy VARCHAR(255) NOT NULL,
        data DATE NOT NULL,
        ora TIME NOT NULL,
        risultato VARCHAR(50),
        FOREIGN KEY (team_alt) REFERENCES team(nome),
        FOREIGN KEY (team_enemy) REFERENCES team(nome)
      )
    `);
    console.log('Tabelle team e partita resettate');

    // Lettura e popolamento da data.tsv (file nella cartella superiore a src/)
    const dataPath = path.join(__dirname, '..', 'data.tsv');
    const file = fs.readFileSync(dataPath, 'utf-8');
    const lines = file.trim().split('\n');
    lines.shift(); // rimuove intestazione

    for (const line of lines) {
      const [team_alt, team_enemy, dataPartita, oraPartita, risultatoRaw] = line.split('\t');
      const [gg, mm, aaaa] = dataPartita.split('/');
      const dataISO = `${aaaa}-${mm.padStart(2, '0')}-${gg.padStart(2, '0')}`;

      // Inserimento dei team
      await conn.query('INSERT IGNORE INTO team(nome) VALUES(?)', [team_alt]);
      await conn.query('INSERT IGNORE INTO team(nome) VALUES(?)', [team_enemy]);

      // Inserimento della partita
      await conn.query(
        `INSERT INTO partita (team_alt, team_enemy, data, ora, risultato)
         VALUES (?, ?, ?, ?, ?)`,
        [team_alt, team_enemy, dataISO, oraPartita, risultatoRaw || null]
      );
    }
    console.log('Database popolato da data.tsv');
  } catch (err) {
    console.error("Errore durante l'inizializzazione del DB:", err);
  } finally {
    if (conn) conn.release();
  }
}

// Endpoint per restituire le partite
app.get('/partite', async (req, res) => {
  let conn;
  try {
    conn = await pool.getConnection();
    const rows = await conn.query(
      `SELECT team_alt, team_enemy, data, ora, risultato
       FROM partita
       ORDER BY data, ora`
    );

    const result = rows.map(r => ({
      team_alt: r.team_alt,
      team_enemy: r.team_enemy,
      data: r.data.toISOString().slice(0, 10),
      ora: r.ora.slice(0, 5),
      risultato: r.risultato
    }));

    res.json(result);
  } catch (err) {
    console.error('Errore in /partite:', err);
    res.status(500).json({ errore: err.message });
  } finally {
    if (conn) conn.release();
  }
});

// Avvio del server
const PORT = 8000;
app.listen(PORT, () => {
  console.log(`Server in ascolto sulla porta ${PORT}`);
  initDb();
});
